WT-12367 Implement the main test program for the model

pmacko86 opened the pull request.

korteland commented on test/model/tools/model_test/main.cpp:
```diff
@@ -0,0 +1,379 @@
+int
+main(int argc, char *argv[])
+{
+    model::kv_workload_generator_spec spec;
```
I can't remember sorry - what does the style guide say about declaring before use? It feels a bit outdated/pointless in C++-land, but I can understand if it's required for some consistency with the C parts of the codebase.

pmacko86 replied:
Fixed; there is no reason that all these variables should be declared before initialization. A power of habit.

korteland commented on test/model/tools/model_test/main.cpp:
```diff
@@ -0,0 +1,379 @@
+            usage(progname);
+            return EXIT_FAILURE;
+        }
+    } catch (std::exception &e) {
```
A lot of catches in this function feel unnecessary, too (and also won't catch a `model_failure`, I think?)

pmacko86 replied:
Why do you find most of these catches unnecessary? They all surround functions and blocks of code that can throw exceptions. I could, in theory, use one large try-catch block around bigger chunks of code, but I much prefer this approach, so that it is clearer from the error message what failed.

And yes, this would catch model failures, which throw instances of `model::verify_exception`.

korteland replied:
Sorry if that was unclear. My proposal was to not have a try/catch around them at all, and just let the failure do the default thing (print the message, then abort the program and possibly dump core).

pmacko86 replied:
I guess I don't really see a utility in the program aborting and would rather have a clean error message, and also no Coverity issues (this is one of the things on which Coverity likes to pick up).

There are cases where having a core is useful for debugging, but I'm not sure how much this is the case here, as recovery is already long done by the time we would notice a verification error.

korteland commented on test/model/tools/model_test/main.cpp:
```diff
@@ -0,0 +1,379 @@
+        /* Clean up the previous database directory, if it exists. */
+        testutil_remove(home.c_str());
+
+        /* Run and verify the workload. */
```
This comment can go IMO - it's not adding much :-)

pmacko86 replied:
This is a matter of style. In terms of explaining the code below, it indeed does not add much. The reason why it is there is because it acts as a "title" for the code below, so that if you just read those "titles", you will see everything that is happening in the code block without even reading the code.

korteland replied:
Fair enough. I have a pathological thing against comments that repeat the code but you're right that it's a style thing. Feel free to leave it.

pmacko86 commented on test/model/tools/model_test/main.cpp:
```diff
@@ -0,0 +1,379 @@
+            continue;
+        update_spec(spec, conn_config, table_config, line.c_str());
+    }
+    if (!f.eof() && !f.good())
```
No, this actually wouldn't work. Counterintuitively, if `std::getline()` does not process any bytes, it sets the fail bit on the input stream, even if the reason was that it reached EOF. So even at the end of a successful file read, we will end up with both fail bit and eof bit.

korteland replied:
Ouch - I didn't know that. Thanks for the correction.

pmacko86 commented on test/model/tools/model_test/main.cpp:
```diff
@@ -0,0 +1,379 @@
+ */
+static void
+run_and_verify(std::shared_ptr<model::kv_workload> workload, const std::string &home,
+  const std::string &conn_config, std::string &table_config)
```
Fixed.

pmacko86 commented on test/model/src/include/model/driver/kv_workload.h:
```diff
@@ -537,7 +537,8 @@ class kv_workload {
      *     Run the workload in WiredTiger.
      */
-    void run_in_wiredtiger(const char *home, const char *connection_config) const;
+    void run_in_wiredtiger(const char *home, const char *connection_config = nullptr,
```
The function `verify_workload` calls `run_in_wiredtiger` without the last argument. While I was adding the new argument, I decided that I may as well make the other one optional. But I can take that out if need be.

korteland replied:
Up to you - I didn't see that existing usage since it wasn't in the diff. If there's a reason for the arguments to be optional then I'm fine to leave them as-is :-)

pmacko86 commented on test/model/src/include/model/util.h:
```diff
@@ -278,6 +278,19 @@ class config_map {
+     *     Get the corresponding float value. Throw an exception on error.
+     */
+    inline float
+    get_float(const char *key) const
```
This is used in `update_spec` in `main.cpp`. It is not written directly, but it is generated by the preprocessor from some of the macro expansions used in the function.

korteland replied:
Thanks! The fact that it's auto-generated makes me feel a little less bad about missing it ;-)

pmacko86 commented on test/model/test/common/util.cpp:
```diff
@@ -68,6 +69,84 @@ create_tmp_file(const char *dir, const char *prefix, const char *suffix)
+    }
+
+    char *p = nullptr;
+    uint64_t r = (uint64_t)strtoull(str, &p, hex ? 16 : 10);
```
Excellent! I had no idea that this behavior was supported. I will fix this.

pmacko86 commented on test/evergreen.yml:
```diff
@@ -3620,6 +3620,22 @@ tasks:
+    tags: ["model_checking"]
+    commands:
+      - func: "get project"
+      - func: "compile wiredtiger"
```
There is already `-DENABLE_MODEL`, see `test/CMakeLists.txt`. It is just not yet used in `evergreen.yml`, because there wasn't a need to disable it yet.

korteland replied:
Cool, thanks for pointing that out.

pmacko86 commented:
I just pushed a few changes, which address the issue above that I mentioned in the comments.

I also made a few more changes:
- Fixed a bug with using the seed via the command-line argument due to iterations being 1-based.
- Removed a check for the `crash_index` advancing from the `kv_workload_runner_wt`, which was creating false positives when the workload generated two `crash` operations back to back. I just removed it because I didn't feel like the check was really necessary.
- Added Evergreen task `model-test-long` also to the ARM platform.

korteland reviewed (approved):
LGTM!

tod-johnson-mongodb commented on test/model/src/include/model/util.h:
```diff
@@ -278,6 +278,19 @@ class config_map {
+    inline float
+    get_float(const char *key) const
+    {
+        std::istringstream stream(std::get<std::string>(_map.find(key)->second));
```
A more efficient way: https://en.cppreference.com/w/cpp/string/basic_string/stof

      return std::stof(_map.find(key)->second);

tod-johnson-mongodb commented on test/model/src/include/model/util.h:
```diff
@@ -304,6 +317,19 @@ class config_map {
+    keys() const noexcept
+    {
+        std::vector<std::string> r;
+        for (auto p : _map)
```
auto makes code harder to read. If the type is not convoluted it is better to use that instead of auto.

pmacko86 replied:
Good point—sometimes I get lazy with `auto` :) Fixed in [ca4ce00](https://github.com/wiredtiger/wiredtiger/pull/10251/commits/ca4ce0065a9940d3d15c15c36ee8877c7bedce9c).
